synapse/storage/databases/main/events_worker.py

```diff
@@ -736,35 +736,80 @@ async def get_stripped_room_state_from_event_context(
+            else:
+                should_start = False
+
+        async def _do_fetch() -> None:
+            """Services requests for events from the `_event_fetch_list` queue."""
+            exc = None
+            try:
+                await self.db_pool.runWithConnection(self._do_fetch)
+            except Exception as e:
+                exc = e
+                raise
+            finally:
+                should_restart = False
+                failed_event_list = []
+                with self._event_fetch_lock:
+                    self._event_fetch_ongoing -= 1
+                    event_fetch_ongoing_gauge.set(self._event_fetch_ongoing)
+
+                    if self._event_fetch_ongoing == 0 and self._event_fetch_list:
+                        # We are the last remaining fetcher and we are about to
+                        # go away. Deal with any outstanding fetches.
+                        if exc is None:
+                            should_restart = True
+                        else:
+                            failed_event_list = self._event_fetch_list
+                            self._event_fetch_list = []
+
+                if should_restart:
+                    self._start_fetch_thread()
+
+                if exc is not None:
+                    for _, deferred in failed_event_list:
+                        if not deferred.called:
+                            with PreserveLoggingContext():
+                                deferred.errback(exc)
+
+        if should_start:
+            run_as_background_process("fetch_events", _do_fetch)
+
     def _do_fetch(self, conn: Connection) -> None:
         """Takes a database connection and waits for requests for events from
         the _event_fetch_list queue.
         """
-        try:
-            i = 0
-            while True:
-                with self._event_fetch_lock:
-                    event_list = self._event_fetch_list
-                    self._event_fetch_list = []
-
-                    if not event_list:
-                        single_threaded = self.database_engine.single_threaded
-                        if (
-                            not self.USE_DEDICATED_DB_THREADS_FOR_EVENT_FETCHING
-                            or single_threaded
-                            or i > EVENT_QUEUE_ITERATIONS
-                        ):
-                            break
-                        else:
-                            self._event_fetch_lock.wait(EVENT_QUEUE_TIMEOUT_S)
-                            i += 1
-                            continue
-                    i = 0
+        i = 0
+        while True:
+            with self._event_fetch_lock:
+                event_list = self._event_fetch_list
+                self._event_fetch_list = []
+
+                if not event_list:
+                    single_threaded = self.database_engine.single_threaded
+                    if (
+                        not self.USE_DEDICATED_DB_THREADS_FOR_EVENT_FETCHING
```

Comment:
I think `USE_DEDICATED_DB_THREADS_FOR_EVENT_FETCHING` might have outlived its purpose. Certainly it no longer seems to match the comment. I'd be interested to know if something has changed here, and whether we should update the comment or get rid of it entirely.

I'm a bit concerned its presence invalidates your carefully constructed test case.

Reply:
I don't think much has changed [since it was introduced](https://github.com/matrix-org/synapse/pull/8450). It looks like a flag to prevent the idling behaviour during tests, since database transactions get run on the main thread.

Reply:
Neither of the two new tests test the threading behaviour deeply enough for it to make a difference

Reply:
fair enough. I'm still giving it side-eye, but let's leave it alone for now.